runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/util/BoundedQueueExecutor.java

```diff
@@ -51,7 +55,31 @@ public BoundedQueueExecutor(
             keepAliveTime,
             unit,
             new LinkedBlockingQueue<>(),
-            threadFactory);
+            threadFactory) {
+          @Override
+          protected void beforeExecute(Thread t, Runnable r) {
+            super.beforeExecute(t, r);
+            synchronized (this) {
+              if (activeCount.get() == maximumPoolSize - 1) {
+                startTimeMaxActiveThreadsUsed = System.currentTimeMillis();
+              }
+              activeCount.incrementAndGet();
+            }
+          }
+
+          @Override
+          protected void afterExecute(Runnable r, Throwable t) {
+            super.afterExecute(r, t);
+            synchronized (this) {
+              if (activeCount.get() == maximumPoolSize) {
```

Comment:
nit Same idea, getAndDecrement()... I guess you'd just need the synchronized block on the long